utils/kafka-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/exception/KafkaJsonExceptionSerializationSchema.scala

```diff
@@ -14,11 +16,123 @@ import java.io.{PrintWriter, StringWriter}
 import java.lang
 import java.net.InetAddress
 import java.nio.charset.StandardCharsets
+import scala.annotation.tailrec
 import scala.io.Source
 
+object KafkaJsonExceptionSerializationSchema {
+  private[exception] val recordOverhead = DefaultRecordBatch.RECORD_BATCH_OVERHEAD
+  private val indentLength              = 2
+  private val warningKey                = "!warning"
+
+  private def serializeValueWithSizeLimit(value: KafkaExceptionInfo, maxValueBytes: Int): Array[Byte] = {
+    val valueJson       = value.asJson
+    val serializedValue = valueJson.spaces2.getBytes(StandardCharsets.UTF_8)
+    if (value.inputEvent.isDefined && serializedValue.length > maxValueBytes) {
+      removeInputEventKeys(valueJson, serializedValue.length, maxValueBytes).spaces2.getBytes(StandardCharsets.UTF_8)
+    } else {
+      serializedValue
+    }
+  }
+
+  private def removeInputEventKeys(value: Json, valueBytes: Int, maxValueBytes: Int): Json = {
+    if (valueBytes <= maxValueBytes) {
+      return value
+    }
+
+    val removedBytesPlaceholder = "$" * valueBytes.toString.length
```

Comment:
You use valueBytes length for the placeholder based on which you compute bytesToCut, and then you replace this placeholder with removedBytes length which is different.  Also, you don't consider the size of the variable keys (I know why, but it wasn't straightforward when I read it for the first time). I see some hidden assumptions here - please document them in the comment.

Reply:
Explanation added, for both